### ISSUE TITLE
Module Request: Blackmagic Design Teranex AV

### ISSUE THREAD

**Issue by rcharters** [bitfocus/companion#599]
**Describe the feature**
Module to control Blackmagic Design Teranex Products, specifically the Teranex AV.
Teranex AV is a video format converter with many other functions, such as color control, audio embedding, timecode control, closed captioning etc.

Initially commands to control Preset Groups in the Teranex AV would be most useful.

**Is this platform dependent (windows, mac, ..)?**
Not Platform Dependent

**If documentation is required to implement, do you know where to find it?**
Protocol to control the Teranex AV is TCP based, and is similar to other Blackmagic Design products such as the Videohub and ATEM products. There is a protocol in the Blackmagic Design Manual for the Teranex Products: https://documents.blackmagicdesign.com/UserManuals/TeranexManual.pdf

**Usecases**
When using multiple Teranex AVs in a production environment, the ability to remotely control the units is desired. Blackmagic Design has a software environment for this purpose, but it is very limited.

Teranex AVs can be used in many instances, such as an interface from a computer to a projector. Converting 4K to HD or HD to 4K. Embedding audio to an HDI SDI Stream. Plus many many more.

**Comment by haakonnessjoen** [comment]
This has already been started on. If you download latest developer release, it has currently only support for changing inputs on audio and video. But format-selection, etc. will come soon.

**Comment by rcharters** [comment]
> This has already been started on. If you download latest developer release, it has currently only support for changing inputs on audio and video. But format-selection, etc. will come soon.

Excellent news - sorry for the repeat, I did not see it listed as a feature but have just found the Support being added soon section. Any idea on the timeframe for this feature? Is there any way I can contribute?

**Comment by haakonnessjoen** [comment]
Which features do you need to control the most, in the Teranex?

**Comment by rcharters** [comment]
> Which features do you need to control the most, in the Teranex?

The ability to Recall Presets (1-6) for the Teranex AV would allow for many options until full functionality is added.

**Comment by chrisrouse** [comment]
Same request for me. We're controlling multiple Teranex units, but we don't mind duplicating the module how many times we need rather than supporting multiple IPs in the same module.

**Comment by chrisrouse** [comment]
I found the commands in the manual (hadn't scrolled are enough before). I've roughed out some code for presets that I'll be able to test tomorrow.

**Comment by chrisrouse** [comment]
@rcharters I pushed an update this morning that was merged that allows selecting the presets. Looks like it works regardless of setting custom names or name for presets.

**Comment by chrisrouse** [comment]
I've added two more features for selecting test patterns and selecting the output display. Are there any other features that would make sense to be controlled from Companion? Pretty much everything else seems like it 's better to control from the Teranex software or from the panel itself because they are more for fine tuning settings.

**Comment by chrisrouse** [comment]
This issue can be closed and marked as complete.

**Comment by fraktalDS** [comment]
Hi chrisrouse,

The SDI input command doesn't work. I guess the problem is that Teranex has 2 SDI inputs and in the command setup I can choose only a simple SDI. Probably you have to add to the command wich SDI input you want to choose. 

Thanks

**Comment by chrisrouse** [comment]
There's no reason it shouldn't work. There's only one SDI input option available on the front panel and in the programming instructions in the manual. There's no option from that side to specify. According to the manual, pressing the SDI button twice changes the input of the SDI button. I didn't test with a connected input, but in my testing while building the input section, I was able to switch between inputs and saw the buttons and screen labels change.

> On Teranex AV, pressing the SDI button more than once will alternate between the two rear panel SDI inputs, SDI 1 and SDI 2. On the first button press, the LCD will momentarily display which SDI input is currently selected. If you press SDI again within 5 seconds, the Teranex AV will switch to your other SDI input signal.

<img width="1257" alt="Screen Shot 2019-05-08 at 7 48 23 PM" src="https://user-images.githubusercontent.com/2112469/57421791-65659780-71ca-11e9-897c-bb0ce9f1f4e5.png">

**Comment by fraktalDS** [comment]
We are talking about this TeranexAV: https://www.blackmagicdesign.com/products/teranex ?
Because there is no component or composite input. I have an AV and I tried your module and all functions are working well except switch to SDI input. The AV has 2 separate SDI inputs, that's why I think you have to choose somehow which one you want. You're right you can switch those 2 SDIs if you push twice the button.
If you think I would happy if I can test for you if you add (for example) a number to the end of the command.

**Comment by willosof** [comment]
TL;DR: https://www.blackmagicdesign.com/products/teranex/techspecs

Two models with different features. 
* Teranex AV
* Teranex Express.

**Comment by fraktalDS** [comment]
I thought. In this case (FYI) this module is working with AV as well, expect the SDI input. Have you any idea?

**Comment by chrisrouse** [comment]
I’m having someone else test it for me since I don’t have access to the hardware, but I don’t see any other options in the documentation to change inputs. What’s there is what I built into the module.

I’ll send you instructions later for testing with telnet to see if the results are any different.

Thanks,
Chris Rouse

**Comment by fraktalDS** [comment]
Thanks Chris!

I'm not sure to have I telnet, but I'll try to do my best :)

**Comment by chrisrouse** [comment]
Are you on Mac or Windows?

Thanks,
Chris Rouse

**Comment by fraktalDS** [comment]
Mac

**Comment by chrisrouse** [comment]
Enable Telnet Mac: http://osxdaily.com/2018/07/18/get-telnet-macos/

Once you have it installed, you'll need to run a couple of commands. Hopefully I'm formatting these correctly since I'm doing this with just the manual in a hotel room.

Connect to the Teranex: `telnet ###.###.###.### 9800` where the ### are the device IP address for the Teranex. I'll leave it up to you to configure the the IP correctly. Press return and you should connect to the Teranex and get a bunch of data in return to give you the current status o the device. 

Once you've got that working, you can send a command to change inputs. Type each line exactly as it is formatted below. Press return ONCE after each line and TWICE after the last line to send the command.

`VIDEO INPUT:` (press return once)
`Video source: SDI` (press return twice)

You could also try sending

`VIDEO INPUT:` (press return once)
`Video mode:` (press return twice)

which should tell you the name of the selected input to see if SDI A or SDI B give different values.

**Comment by haakonnessjoen** [comment]
@fraktalDS when you have telnet working, you don't really need to test sending the commands that @chrisrouse mentions.

The easiest way to test, is that you connect to the device as @chrisrouse said. But after telnet is connected (you will get a lot of text), try to push the sdi button twice on your front panel, and be sure to hit HDMI too. Just to get the different values. If you glance at the telnet window, you should se three "VIDEO INPUT:" sections. Please copy and paste the data you get, here :)

**Comment by fraktalDS** [comment]
HI guys,

I've really appreciated your help! I've already installed the telnet and it seems working well. I'm on an event until Monday. I'll come back on Monday if I could test it.

Many thanks,
Zoltan

**Comment by haakonnessjoen** [comment]
Please copy and paste the output you get when you click several times on the sdi button, when you have time! :)

**Comment by fraktalDS** [comment]
Hi guys,

First of all, I would like to say thank you for your fast help! You are awesome! :)
Secondly, here you are the answer from the AV:

once:

VIDEO INPUT:
Video source: SDI1
Audio source: Embedded

twice:

VIDEO INPUT:
Video source: SDI1
Audio source: Embedded

VIDEO INPUT:
Video source: SDI2
Audio source: Embedded

If it is interest for you here you are the response if the input present:

VIDEO INPUT:
Video mode: 2160p50
Signal present: true

VIDEO ADJUST:
Luma low: 4
Luma high: 1019
Chroma low: 4
Chroma high: 1019
Aspect fill luma: 64
Aspect fill Cb: 512
Aspect fill Cr: 512

VARIABLE ASPECT RATIO:
Variable Aspect Ratio size X left: 0.000000
Variable Aspect Ratio size X right: 0.000000
Variable Aspect Ratio size Y left: 0.000000
Variable Aspect Ratio size Y right: 0.000000
Variable Aspect Ratio pos X left: 0.000000
Variable Aspect Ratio pos X right: 0.000000
Variable Aspect Ratio pos Y left: 0.000000
Variable Aspect Ratio pos Y right: 0.000000
Variable Aspect Ratio trim X left: 0
Variable Aspect Ratio trim X right: 0
Variable Aspect Ratio trim Y left: 0
Variable Aspect Ratio trim Y right: 0
Variable Aspect Ratio zoom/crop: false

And one more thing, what would be really great. If you could, can you add a prompt to the screen capture?
Here you are the code:

VIDEO OUTPUT:
Still frame present: true
Still store complete: false

VIDEO OUTPUT:
Still store complete: true

There were about 10s between the 2 messages, and there was a warning message:
"This action will overwrite any existing still at this resolution. 
/Cancel/ /Capture/ "
I guess that's why the first complete: false respond.

If I can do anything else, just give me a shout!

Many thanks,
Zoltan

**Comment by chrisrouse** [comment]
How about that! I’ll push an update this morning.

Thanks,
Chris Rouse

**Comment by fraktalDS** [comment]
Brilliant!

I've sent the command what you suggested. Here you are this part:

VIDEO INPUT:    
Video source: SDI1

ACK

VIDEO INPUT:
Video source: SDI1
Audio source: Embedded
Signal present: false

**Comment by chrisrouse** [comment]
If you have just a single SDI input connected does it return SDI 1 and SDI 2 still? Just want to make sure it’s consistent.

Thanks,
Chris Rouse

**Comment by fraktalDS** [comment]
Yes.

Here is the whole response, if the SDI1 signal present (I sent the command from telnet):
VIDEO INPUT:
Video source: SDI1

ACK

VIDEO INPUT:
Video source: SDI1
Audio source: Embedded

VIDEO INPUT:
Signal present: false

VIDEO INPUT:
Video mode: 2160p30
Signal present: true

VIDEO ADJUST:
Luma low: 4
Luma high: 1019
Chroma low: 4
Chroma high: 1019
Aspect fill luma: 64
Aspect fill Cb: 512
Aspect fill Cr: 512

VARIABLE ASPECT RATIO:
Variable Aspect Ratio size X left: 0.000000
Variable Aspect Ratio size X right: 0.000000
Variable Aspect Ratio size Y left: 0.000000
Variable Aspect Ratio size Y right: 0.000000
Variable Aspect Ratio pos X left: 0.000000
Variable Aspect Ratio pos X right: 0.000000
Variable Aspect Ratio pos Y left: 0.000000
Variable Aspect Ratio pos Y right: 0.000000
Variable Aspect Ratio trim X left: 0
Variable Aspect Ratio trim X right: 0
Variable Aspect Ratio trim Y left: 0
Variable Aspect Ratio trim Y right: 0
Variable Aspect Ratio zoom/crop: false

**Comment by fraktalDS** [comment]
It seems the capture won't be easy. I can't find any command in the user manual to CAPTURE, just to ask the state of the still store:

Still frame present: | For Teranex AV only. True = still frame has been stored at the current output format. False = no still frame has been stored at the current output format.Cannot be changed by the user.
Still load complete: | For Teranex AV only. True = loading of still frame data from the internal non-volatile flash drive has completed.Cannot be changed by the user.
Still store complete: | For Teranex AV only. True = transfer of still frame data to the internal non-volatile flash drive has completed.Cannot be changed by the user.

If I capture a still frame from the software or the front panel the telnet says:

VIDEO OUTPUT:
Still frame present: true
Still store complete: false

VIDEO OUTPUT:
Still store complete: true

with a couple of seconds between the 2 video output lines (I don't do anything during this time).
But if send any combination of these lines (with or without "true/false", one or two lines, exactly the same as up), nothing happens.
Any ideas?

**Comment by chrisrouse** [comment]
Pull request added to included an SDI 1 and SDI 2 option. https://github.com/bitfocus/companion-module-bmd-teranex/pull/3

**Comment by chrisrouse** [comment]
Still store can only be initiated from the software on the computer. The data provided is simply feedback. It would be amazing to do it from a single button, especially when you're using multiple Teranex devices at the same time with different inputs, but unfortunately I don't think that's possible without an update from Blackmagic.

**Comment by josephdadams** [comment]
Moving this issue into the Teranex repo. You guys can handle it from there, if you want to keep it open, rename, close, whatever floats your boat.
